bst.py

```diff
@@ -0,0 +1,120 @@
+            else:
+                self.right.insert(val)
+
+    def contains(self, val):
+        """Check tree for node with given value.
+
+        args:
+            val: the value to check for
+
+        returns: True if val is in the tree, False if not.
+        """
+        if val == self.val:
+            return True
+        elif val < self.val:
+            if self.left is None:
+                return False
+            return self.left.contains(val)
+        elif val > self.val:
+            if self.right is None:
+                return False
+            return self.right.contains(val)
+
+    def size(self):
+        """Return the total number of nodes in the tree.
+
+        returns: integer of total node; 0 if empty
+        """
+        left_size = self.left.size() if self.left is not None else 0
+        right_size = self.right.size() if self.right is not None else 0
+        return left_size + right_size + 1
+
+    def depth(self):
+        """Return an the total number of levels in the tree.
+
+        If there is one value, the depth should be 1, if two values it'll be 2,
+        if three values it may be 2 or three, depending, etc.
+
+        returns: integer of level number
+        """
+        left_depth = self.left.depth() if self.left is not None else 0
+        right_depth = self.right.depth() if self.right is not None else 0
+        return max(left_depth, right_depth) + 1
+
+    def balance(self):
+        """Return a positive or negative number representing tree balance.
+
+        Trees higher on the left than the right should return a positive value,
+        trees higher on the right than the left should return a negative value.
+        An ideally-balanced tree should return 0.
+
+        returns: integer
+        """
+        left_depth = self.left.depth() if self.left is not None else 0
+        right_depth = self.right.depth() if self.right is not None else 0
+        if left_depth > right_depth:
+            return 1
+        elif left_depth < right_depth:
+            return -1
+        else:
+            return 0
+
+
+if __name__ == '__main__':
+    from timeit import Timer
+
+    """Document the best and worst cases for searching for a value in the tree.
+        The worst case consists of a tree with one long linear branch.
+        The best case is a perfectly balanced tree.
+    """
+    worst = Node(1)
+    for val in range(2, 32):
+        worst.insert(val)
+
+    best = Node(16)
+    best_values = [
+        8, 24, 4, 12, 20, 28, 2, 6, 10, 14, 18, 22, 26, 30, 1, 3, 5,
+        7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31
+    ]
+    for val in best_values:
+        best.insert(val)
+
+    worst_case = Timer(
+        'worst.contains(31)', 'from __main__ import worst').timeit(1000)
+    best_case = Timer(
+        'best.contains(31)', 'from __main__ import best').timeit(1000)
+
+    print "The worst case took {}.".format(worst_case)
+    print "The best case took {}.".format(best_case)
```

Comment:
given the small size of your trees, I don't imagine this difference is very large.